proposals/0000-custom-type-errors.rst

```diff
@@ -0,0 +1,188 @@
+.. proposal-number:: 
+
+.. trac-ticket:: 
+
+.. implemented:: 
+
+.. highlight:: haskell
+
+This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/59>`_.
+
+.. contents::
+
+Custom type errors
+==================
+
+This proposes the addition of two combinators over ``Constraint``, namely ::
+
+    IfUndischarged :: Constraint -> ErrorMessage -> Constraint
+    IfApart        :: Type -> Type -> Constraint -> Constraint -> Constraint
+
+which enable library authors to provide custom error messages for their libraries.
+
+In addition, this proposal also motivates changing the constraint solving order to prioritize arguments to functions over the check of the application itself.
+
+This proposal does *not* address whether this facility should be used at all by the libraries bundled with GHC. This might be the topic of another proposal in the future, if this one gets accepted.
+
+Further information (with a more academic flavor) is available in `this draft <http://www.staff.science.uu.nl/~f100183/type-errors-draft.pdf>`_ and this `presentation at Curry On! 2017 <https://www.youtube.com/watch?v=LuqSkWOcnSA>`_.
+
+
+Motivation
+------------
+Haskell is well-known for its ability to embed DSLs, and the usage of its strong type system to guarantee invariants in those languages. The main disdvantage is that type errors become unwidely long and complex, and in many cases expose internals of the implementation. In short, the "ultimate abstraction" offered by EDSLs is broken when the code is ill-typed.
+
+The community has already acknowledged this problem partially with the introduction of ``TypeError``. See `the original proposal <https://ghc.haskell.org/trac/ghc/wiki/Proposal/CustomTypeErrors>`_ and an `example package <https://github.com/turingjump/bookkeeper#readme>`_. But why should we stop here? There are many other error conditions we may want to customize. In this proposal we focus in other two.
+
+Undischarged constraints
+~~~~~~~~~~~~~~~~~~~~~~~~
+These are error messages which are shown when a given constraint is left undischarged at the end of the solving process. The archetypal example is a type class instance constraint which is not satisfies. For example, if it turns out that ``(==)`` is applied to a type for which no ``Eq`` instance exists, the error message could become ::
+
+  * X does not implement equality.
+    Maybe you want to derive it using `deriving Eq`?
+  * In the expression: ...
+
+We definitely could include a generic framework to hint auto-derivation of built-in classes such as ``Eq`` or ``Ord``. But in the current Haskell ecosystem some libraries use generic deriving, others Template Haskell, and so on, so a generic facility seems useful.
+
+Inequalities
```

Comment:
The problem is explained in Section 4 of the draft; his version of `IfUndischarged` for equality constraints is similar to the `InconsistentHint` there. In short, adding an error message without an underlying structure of how constraints are solved leads to error messages being reported for the wrong constraints, and thus to wrong error messages.